File: python-leetcode/array/240.搜索二维矩阵 II.py
from typing import List


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        # 对每一层进行二分查找
        # 时间复杂度 O(mlogn)
        for nums in matrix:
            idx = self.search(nums, target)
            if idx != -1:
                return True
        return False

    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1

        while left <= right:
            mid = left + (right - left) // 2
            if target > nums[mid]:
                left = mid + 1
            elif target < nums[mid]:
                right = mid - 1
            else:
                return mid
        return -1

File: go-leetcode/array/240.搜索二维矩阵 II.go
package leetcode

func searchMatrix(matrix [][]int, target int) bool {
	// 对每一层进行二分查找
	// 时间复杂度 O(mlogn)
	for _, nums := range matrix {
		idx := search(nums, target)
		if idx != -1 {
			return true
		}
	}
	return false
}

func search(nums []int, target int) int {
	left := 0
	right := len(nums) - 1

	for left <= right {
		mid := left + (right-left)/2
		if target > nums[mid] {
			left = mid + 1
		} else if target < nums[mid] {
			right = mid - 1
		} else {
			return mid
		}
	}
	return -1
}
